### PR TITLE
Terms sense

### PR DESCRIPTION
**Pull request by szymon0w**
Created new table search_phrase thanks to it one category can have multiple search phrases and each search phrase can have different number of words downloaded.

### CODE REVIEW

File: thesis-api/main/database/models.py [hunk 1]
```diff
@@ -11,6 +11,7 @@ class CategoryModel(Base):
     id = Column(Integer, primary_key=True, nullable=False)
     name = Column(String, unique=True, nullable=False)
     search_word = Column(String, unique=True, nullable=False)
+    no_of_records = Column(Integer, unique=False, nullable=False)
```
Comment (Lozoko) [inline thesis-api/main/database/models.py:14]: Unnecessary change, we do not use thesis-api anymore.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SearchPhraseService.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import pl.edu.agh.quizzesthesis.data.repository.SearchPhraseRepository;
+
+import java.util.List;
+
+@Service
+@AllArgsConstructor
+public class SearchPhraseService {
+
+    private final SearchPhraseRepository searchPhraseRepository;
+    public List<SearchPhrase> getAll() {
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SearchPhraseService.java:16]: Missing new line.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java [hunk 3]
```diff
@@ -28,8 +30,10 @@
 public class SetupService {
 
     private static final String CATEGORIES_FILE_PATH = "categories.txt";
-
+    private static final String SEARCH_PHRASES_FILE_PATH = "search_phrases.txt";
     private final CategoryRepository categoryRepository;
+    private final SearchPhraseRepository searchPhraseRepository;
+
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java]: Unnecessary empty line.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java [hunk 4]
```diff
@@ -38,18 +42,23 @@ public class SetupService {
...
         CSVParser parser = new CSVParserBuilder().withSeparator(';').build();
         try (
-
                 var reader = new BufferedReader(new InputStreamReader(new ClassPathResource(CATEGORIES_FILE_PATH).getInputStream()));
                 CSVReader csvReader = new CSVReaderBuilder(reader)
                         .withSkipLines(1)
                         .withCSVParser(parser)
-                        .build()
+                        .build();
+                var reader2 = new BufferedReader(new InputStreamReader(new ClassPathResource(SEARCH_PHRASES_FILE_PATH).getInputStream()));
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java]: We can't have some ambiguous `reader2`, rename it to something that indicates its purpose, for example `searchPhraseReader`. Also, it would be nice to rename the previous reader too.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java [hunk 5]
```diff
@@ -38,18 +42,23 @@ public class SetupService {
...
+                CSVReader csvReader2 = new CSVReaderBuilder(reader2)
+                .withSkipLines(1)
+                .withCSVParser(parser)
+                .build()
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java]: Missing indentation.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java [hunk 6]
```diff
@@ -68,13 +77,37 @@ public void setUp() {
             categoryRepository.saveAll(categoriesToUpdate);
             var categoriesPersisted = categoryRepository.saveAll(categoriesToPersist);
 
+            var allCategories = categoryRepository.findAll().stream()
+                    .collect(Collectors.toMap(Category::getName, category -> category));
+
+            var searchPhrasesInFile = csvReader2.readAll().stream()
+                    .map(searchPhraseRow -> new SearchPhrase(null,searchPhraseRow[1],Integer.valueOf(searchPhraseRow[2]), allCategories.get(searchPhraseRow[0])))
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java:82]: This line is way too long, split it for better readability. Also, pro tip, use ctrl + alt + l for automatic style formatting, it will for example add missing space characters after commas ;)

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/TermService.java [hunk 7]
```diff
@@ -82,9 +82,11 @@ public boolean existsOfCategory(String termName, int categoryId) {
         return termRepository.existsByNameAndCategoryId(termName, categoryId);
     }
 
-    private void setupOne(final Category category) {
-        var termsAlreadyPersisted = termRepository.findAllByCategoryId(category.getId());
-        if (termsAlreadyPersisted.size() >= 1000) {
+    private void setupOne(final SearchPhrase searchPhrase) {
+        int noOfRecords = searchPhrase.getNoOfRecords();
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/TermService.java:86]: Use `var` when you are already using it for the rest of variables.

Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/TermService.java:86]: I don't like the name `noofRecords` (especially `no`) but it is my personal opinion and you can ignore it if you disagree with me

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/TermService.java [hunk 8]
```diff
@@ -124,7 +126,7 @@ private void setupOne(final Category category) {
             return;
         }
 
-        newTermsFromApi = newTermsFromApi.subList(0, min(newTermsFromApi.size(), 1000 - termsAlreadyPersisted.size()));
+        newTermsFromApi = newTermsFromApi.subList(0, min(newTermsFromApi.size(), noOfRecords - numberOfTermsBySearchPhrase));
         float maxFrequency = termsFromApi.get(0).frequency();
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/TermService.java:130]: This doesn't have any sense since we do not have any sorting (at least in this version of branch)

Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/TermService.java:130]: Since we have multiple setups for terms from single category, it would be best to set initial difficulty based on new terms from Api **and** terms already persisted in database.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/WrongAnswerService.java [hunk 9]
```diff
@@ -79,7 +79,7 @@ public List<String> getWrongAnswers(Term rightAnswerTerm) {
                     break;
                 }
             }
-            if (!termService.existsOfCategory(potentialAnswer.word(), rightAnswerTerm.getCategory().getId())){
+            if (!termService.existsOfCategory(potentialAnswer.word(), rightAnswerTerm.getSearchPhrase().getCategory().getId())){
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/WrongAnswerService.java:82]: Isn't it the same as before? It would be really helpful if you provide a reason for that change (sorry for my laziness)

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/entity/SearchPhrase.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+public class SearchPhrase {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Integer id;
+    private String searchWord;
+    private Integer noOfRecords;
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/entity/SearchPhrase.java:22]: As before, I don't like this name. Just personal thing.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/entity/SearchPhrase.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+    @ManyToOne
+    private Category category;
+
+
+
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/entity/SearchPhrase.java:28]: Unnecessary empty space.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/SearchPhraseRepository.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+import java.util.List;
+
+public interface SearchPhraseRepository extends CrudRepository<SearchPhrase, Integer> {
+
+    List<SearchPhrase> findAll();
+
+    @Transactional
+    void deleteAllByCategory(Category category);
+}
+
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/SearchPhraseRepository.java:17]: Unnecessary empty line.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/TermRepository.java [hunk 13]
```diff
@@ -4,21 +4,23 @@
...
-import pl.edu.agh.quizzesthesis.data.entity.Category;
+import pl.edu.agh.quizzesthesis.data.entity.SearchPhrase;
 import pl.edu.agh.quizzesthesis.data.entity.Term;
 
 import java.util.Set;
 
 public interface TermRepository extends PagingAndSortingRepository<Term, Integer> {
 
     boolean existsByNameAndCategoryId(String termName, int categoryId);
+    Set<Term> findAllByCategoryId(int searchPhraseId);
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/TermRepository.java:15]: So by `categoryId` or `searchPhraseId`? Decide which one.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/TermRepository.java [hunk 14]
```diff
@@ -4,21 +4,23 @@
...
 
-    Set<Term> findAllByCategoryId(int categoryId);
 
     int countByCategoryId(int categoryId);
+    int countBySearchPhraseId(int categoryId);
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/TermRepository.java:19]: So by `categoryId` or `searchPhraseId`? Decide which one.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/TermRepository.java (hunk 13 above)
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/TermRepository.java:15]: Missing new line character, again ctrl alt l.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/TermRepository.java (hunk 14 above)
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/TermRepository.java:19]: Missing new line character, again ctrl alt l.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/TermRepository.java [hunk 15]
```diff
@@ -4,21 +4,23 @@
...
 
     Page<Term> findPageByCategoryId(PageRequest pageRequest, int categoryId);
 
     @Transactional
-    void deleteAllByCategory(Category category);
+    void deleteAllBySearchPhrase(SearchPhrase searchPhrase);
+
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/data/repository/TermRepository.java:25]: We can clearly see what the lack of tests is leading to. This code either is not working or is working but confuses the reviewer and other developers.

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java [hunk 16]
```diff
@@ -38,18 +42,23 @@ public class SetupService {
...
-
+        var searchPhrasesAlreadyPersisted = searchPhraseRepository.findAll().stream()
+                .collect(Collectors.toMap(SearchPhrase::getSearchWord, searchPhrase -> searchPhrase));
         CSVParser parser = new CSVParserBuilder().withSeparator(';').build();
         try (
-
                 var reader = new BufferedReader(new InputStreamReader(new ClassPathResource(CATEGORIES_FILE_PATH).getInputStream()));
                 CSVReader csvReader = new CSVReaderBuilder(reader)
                         .withSkipLines(1)
                         .withCSVParser(parser)
```
Comment (Lozoko) [inline thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java]: This parser is unnecessary (at least in this form) since `categories.txt` does not have `;` character anymore.

File: thesis-api-v2/src/main/resources/categories.txt [hunk 17]
```diff
@@ -1,4 +1,7 @@
-name;search_word
-animals;types of animals
-herbs;types of herbs
-parts of human body;parts of human body
+name
+animals
+cars
+herbs
+parts of human body
+types of cars
```
Comment (Lozoko) [inline thesis-api-v2/src/main/resources/categories.txt]: `Cars` and `Types of cars`? I don't get it? Why two different categories with the same meaning? Please explain it to me.

File: thesis-api-v2/src/main/resources/search_phrases.txt [hunk 18]
```diff
@@ -0,0 +1,11 @@
+name;search_word;quantity
+animals;types of mammals;100
+animals;types of lizards;100
+animals;types of birds;100
+cars; types of cars;100
+herbs;types of herbs;150
+parts of human body;parts of human body;90
+types of cars;types of cars;150
```
Comment (Lozoko) [inline thesis-api-v2/src/main/resources/search_phrases.txt]: Why we have two identical `search_word` in the file? It is pretty dangerous since we are relying on `search_word` property multiple times (for example as keys for maps). If there is proper reason for it, please explain it to me.

File: thesis-api-v2/src/main/resources/search_phrases.txt [hunk 19]
```diff
@@ -0,0 +1,11 @@
+name;search_word;quantity
+animals;types of mammals;100
+animals;types of lizards;100
+animals;types of birds;100
+cars; types of cars;100
```
Comment (Lozoko) [inline thesis-api-v2/src/main/resources/search_phrases.txt]: Is it intended to have a white character before `search_word`? Sure, it makes so that we can have two semi identical `search_word` for two different categories, but it is confusing and I would like to get rid of this approach. If it is just a typo then... please just remove it.

**Review by Lozoko — CHANGES_REQUESTED** [review]
This brach is based on a really old version of master (it lacks picture questions for example) and because of that I can't properly review your changes. Please rebase this branch using current version of master.
As for general idea of fixing the issue, I like the solution you proposed.

File: thesis-api-v2/src/main/resources/search_phrases.txt (hunk 19 above)
Comment (szymon0w) [reply thesis-api-v2/src/main/resources/search_phrases.txt]: there should not be white character and not two semi identical search_words either, I have to spend more time to make a list of all wanted categories and search phrases, this files will change

File: thesis-api-v2/src/main/resources/search_phrases.txt (hunk 18 above)
Comment (szymon0w) [reply thesis-api-v2/src/main/resources/search_phrases.txt]: Fixed

File: thesis-api-v2/src/main/resources/categories.txt (hunk 17 above)
Comment (szymon0w) [reply thesis-api-v2/src/main/resources/categories.txt]: My mistake, this files are still not in final state

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java (hunk 16 above)
Comment (szymon0w) [reply thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java]: That's right

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java (hunk 4 above)
Comment (szymon0w) [reply thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java]: Done

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java (hunk 5 above)
Comment (szymon0w) [reply thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java]: Ok, tabbed it

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java (hunk 3 above)
Comment (szymon0w) [reply thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SetupService.java]: no longer there

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SearchPhraseService.java (hunk 2 above)
Comment (szymon0w) [reply thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/service/SearchPhraseService.java:16]: Added
